packages/agent/src/fetch_candid.ts

```diff
@@ -0,0 +1,21 @@
+import { Actor, ActorSubclass, Agent } from '.';
+import { IDL } from '@dfinity/candid';
+
+// Common interface for the hidden `__get_candid_interface_tmp_hack` actor method
+const tmpHackInterface: IDL.InterfaceFactory = ({ IDL }) =>
+  IDL.Service({
+    __get_candid_interface_tmp_hack: IDL.Func([], [IDL.Text], ['query']),
+  });
+
+/**
+ * Retrieves the Candid interface for the specified canister..
+ *
+ * @param agent The agent to use for the request (usually an `HttpAgent`)
+ * @param canisterId A string corresponding to the canister ID
+ * @returns Candid source code
+ */
+export async function fetchCandid(agent: Agent, canisterId: string): Promise<string> {
```

Comment:
We need to use this logic: https://github.com/dfinity/candid/blob/master/tools/ui/src/candid.ts#L57

Fetch from metadata first, if it fails, fetch from `_tmp_hack`. All canisters built from recent dfx releases should have the metadata already. But we can never drop `_tmp_hack` because some canisters never upgrade.

Reply:
Makes sense. This function now uses the same logic as in the Candid UI.